File: src/Dpz.Core.Web.Dashboard/Shared/Components/AudioPlayer.razor.cs
using System;
using System.Threading.Tasks;
using Dpz.Core.Web.Dashboard.Models.Dialog;
using Dpz.Core.Web.Dashboard.Service;
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;

namespace Dpz.Core.Web.Dashboard.Shared.Components;

public partial class AudioPlayer(IJSRuntime jsRuntime, IAppDialogService dialogService)
    : ComponentBase,
        IAsyncDisposable
{
    [Parameter]
    [EditorRequired]
    public required string Src { get; set; }

    private ElementReference _audioElement;
    private IJSObjectReference? _jsModule;
    private IJSObjectReference? _jsPlayer;
    private DotNetObjectReference<AudioPlayer>? _objRef;

    private bool _isPlaying;
    private double _currentTime;
    private double _duration;

    private double DisplayTime
    {
        get => _currentTime;
        set => _currentTime = value;
    }

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            try
            {
                _objRef = DotNetObjectReference.Create(this);
                _jsModule = await jsRuntime.InvokeAsync<IJSObjectReference>(
                    "import",
                    "./Shared/Components/AudioPlayer.razor.js"
                );
                _jsPlayer = await _jsModule.InvokeAsync<IJSObjectReference>(
                    "init",
                    _objRef,
                    _audioElement
                );
            }
            catch (Exception ex)
            {
                Console.WriteLine($"AudioPlayer init failed: {ex.Message}");
            }
        }
    }

    private async Task TogglePlay()
    {
        if (_jsPlayer == null)
        {
            return;
        }

        if (_isPlaying)
        {
            try
            {
                await _jsPlayer.InvokeVoidAsync("pause");
            }
            catch (Exception ex)
            {
                await dialogService.ShowAlertAsync(
                    new AppDialogOptions { Title = "错误", Message = $"暂停音频失败：{ex.Message}" }
                );
            }
            _isPlaying = false;
        }
        else
        {
            try
            {
                await _jsPlayer.InvokeVoidAsync("play");
            }
            catch (Exception ex)
            {
                await dialogService.ShowAlertAsync(
                    new AppDialogOptions { Title = "错误", Message = $"播放音频失败：{ex.Message}" }
                );
            }
            _isPlaying = true;
        }
    }

    private async Task OnSeekChange(ChangeEventArgs e)
    {
        if (_jsPlayer == null || !double.TryParse(e.Value?.ToString(), out var time))
        {
            return;
        }

        _currentTime = time;
        try
        {
            await _jsPlayer.InvokeVoidAsync("setCurrentTime", time);
        }
        catch (Exception ex)
        {
            await dialogService.ShowAlertAsync(
                new AppDialogOptions { Title = "错误", Message = $"跳转播放位置失败：{ex.Message}" }
            );
        }
    }

    private void OnSeekInput(ChangeEventArgs e)
    {
        if (double.TryParse(e.Value?.ToString(), out var time))
        {
            _currentTime = time;
        }
    }

    [JSInvokable]
    public void OnTimeUpdate(double time)
    {
        _currentTime = time;
        StateHasChanged();
    }

    [JSInvokable]
    public void OnDurationChange(double duration)
    {
        _duration = duration;
        StateHasChanged();
    }

    [JSInvokable]
    public void OnEnded()
    {
        _isPlaying = false;
        _currentTime = 0;
        StateHasChanged();
    }

    private string FormatTime(double seconds)
    {
        var ts = TimeSpan.FromSeconds(seconds);
        return ts.TotalHours >= 1 ? ts.ToString(@"h\:mm\:ss") : ts.ToString(@"m\:ss");
    }

    public async ValueTask DisposeAsync()
    {
        try
        {
            if (_jsPlayer != null)
            {
                await _jsPlayer.InvokeVoidAsync("dispose");
                await _jsPlayer.DisposeAsync();
            }
            if (_jsModule != null)
            {
                await _jsModule.DisposeAsync();
            }
            _objRef?.Dispose();
        }
        catch
        {
            // ignored
        }
    }
}

File: src/Dpz.Core.Web.Dashboard/Shared/Components/AudioPlayer.razor.js
export class AudioPlayer {
    constructor(dotNetHelper, audioElement) {
        this.dotNetHelper = dotNetHelper;
        this.audio = audioElement;
        this.isDragging = false;
        this.isPlaying = false;
        this.onTimeUpdate = this.handleTimeUpdate.bind(this);
        this.onLoadedMetadata = this.handleLoadedMetadata.bind(this);
        this.onEnded = this.handleEnded.bind(this);
        this.onError = this.handleError.bind(this);
        
        this.setupEvents();
    }

    setupEvents() {
        this.audio.addEventListener("timeupdate", this.onTimeUpdate);
        this.audio.addEventListener("loadedmetadata", this.onLoadedMetadata);
        this.audio.addEventListener("ended", this.onEnded);
        this.audio.addEventListener("error", this.onError);
    }

    handleTimeUpdate() {
        if (!this.audio || !this.dotNetHelper || this.isDragging) {
            return;
        }

        this.dotNetHelper.invokeMethodAsync("OnTimeUpdate", this.audio.currentTime);
    }

    handleLoadedMetadata() {
        if (!this.audio || !this.dotNetHelper) {
            return;
        }

        this.dotNetHelper.invokeMethodAsync("OnDurationChange", this.audio.duration);
    }

    handleEnded() {
        if (!this.dotNetHelper) {
            return;
        }

        this.isPlaying = false;
        this.dotNetHelper.invokeMethodAsync("OnEnded");
    }

    handleError() {
        if (!this.audio || !this.isPlaying) {
            return;
        }
        
        const error = this.audio.error;
        if (!error) {
            return;
        }

        console.warn("Audio playback failed.", {
            code: error.code,
            message: error.message,
            src: this.audio.currentSrc || this.audio.src
        });
    }

    play() {
        this.isPlaying = true;
        return this.audio.play().catch((error) => {
            this.isPlaying = false;
            throw error;
        });
    }

    pause() {
        this.isPlaying = false;
        this.audio.pause();
    }

    setCurrentTime(time) {
        if (isFinite(time)) {
            this.audio.currentTime = time;
        }
    }
    
    setVolume(value) {
        this.audio.volume = value;
    }

    dispose() {
        if (this.audio) {
            this.audio.removeEventListener("timeupdate", this.onTimeUpdate);
            this.audio.removeEventListener("loadedmetadata", this.onLoadedMetadata);
            this.audio.removeEventListener("ended", this.onEnded);
            this.audio.removeEventListener("error", this.onError);
        }

        this.audio = null;
        this.dotNetHelper = null;
    }
}

export function init(dotNetHelper, audioElement) {
    return new AudioPlayer(dotNetHelper, audioElement);
}
